RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -17,58 +21,83 @@
+        return path -> new LuceneDocumentsReader(path, softDeletesPossible, softDeletesField);
+    }
     public static final int NUM_DOCUMENTS_BUFFERED = 1024;
+
     protected final Path indexDirectoryPath;
+    protected final boolean softDeletesPossible;
+    protected final String softDeletesField;
 
     public Flux<Document> readDocuments() {
-        return Flux.using(() -> openIndexReader(indexDirectoryPath), reader -> {
-            log.info(reader.maxDoc() + " documents found in the current Lucene index");
+        return Flux.using(() -> wrapReader(DirectoryReader.open(FSDirectory.open(indexDirectoryPath)), softDeletesPossible, softDeletesField), reader -> {
+            log.atInfo().log(reader.maxDoc() + " documents found in the current Lucene index");
+
+            return Flux.fromIterable(reader.leaves()) // Iterate over each segment
+                .flatMap(leaf -> {
+                    LeafReader leafReader = leaf.reader();
+                    Bits liveDocs = leafReader.getLiveDocs(); // Get live docs bits
+
+                    return Flux.range(0, leafReader.maxDoc())
+                        .handle((docId, sink) -> {
+                            // Check if the document is live (not hard deleted/updated)
+                            boolean isLive = liveDocs == null || liveDocs.get(docId);
+                            Document doc = getDocument(leafReader, docId, isLive);
 
-            return Flux.range(0, reader.maxDoc()) // Extract all the Documents in the IndexReader
-                .handle((i, sink) -> {
-                    Document doc = getDocument(reader, i);
-                    if (doc != null) { // Skip malformed docs
-                        sink.next(doc);
-                    }
+                            if (doc != null) { // Skip malformed docs
+                                sink.next(doc);
+                            }
+                        });
                 })
                 .cast(Document.class);
-        }, reader -> { // Close the IndexReader when done
+        }, reader -> { // Close the DirectoryReader when done
             try {
                 reader.close();
             } catch (IOException e) {
-                log.error("Failed to close IndexReader", e);
+                log.atError().setMessage("Failed to close DirectoryReader").setCause(e).log();
                 throw Lombok.sneakyThrow(e);
             }
         });
     }
 
-    protected IndexReader openIndexReader(Path indexDirectoryPath) throws IOException {
-        return DirectoryReader.open(FSDirectory.open(indexDirectoryPath));
+    protected DirectoryReader wrapReader(DirectoryReader reader, boolean softDeletesEnabled, String softDeletesField) throws IOException {
+        if (softDeletesEnabled) {
+            return new SoftDeletesDirectoryReaderWrapper(reader, softDeletesField);
+        }
+        return reader;
```

Comment:
This needs a comment explaining what it does and how it works; it's pretty magical